nexus/deployment/src/blueprint_builder.rs

```diff
@@ -0,0 +1,682 @@
+        let zones_in_service = blueprint_initial.zones_in_service.clone();
+        let diff = blueprint_initial
+            .diff_from_collection(&collection, &zones_in_service);
+        println!(
+            "collection -> initial blueprint (expected no changes):\n{}",
+            diff
+        );
+        assert_eq!(diff.sleds_added().count(), 0);
+        assert_eq!(diff.sleds_removed().count(), 0);
+        assert_eq!(diff.sleds_changed().count(), 0);
+
+        // Test a no-op blueprint.
+        let builder = BlueprintBuilder::new_based_on(
+            &blueprint_initial,
+            &policy,
+            "test_basic",
+        );
+        let blueprint = builder.build();
+        let diff = blueprint_initial.diff(&blueprint);
+        println!(
+            "initial blueprint -> next blueprint (expected no changes):\n{}",
+            diff
+        );
+        assert_eq!(diff.sleds_added().count(), 0);
+        assert_eq!(diff.sleds_removed().count(), 0);
+        assert_eq!(diff.sleds_changed().count(), 0);
+    }
+
+    #[test]
+    fn test_basic() {
+        let (collection, mut policy) = example();
+        let blueprint1 = BlueprintBuilder::build_initial_from_collection(
+            &collection,
+            &policy,
+            "the_test",
+        )
+        .expect("failed to create initial blueprint");
+
+        let mut builder =
+            BlueprintBuilder::new_based_on(&blueprint1, &policy, "test_basic");
+
+        // The initial blueprint should have internal NTP zones on all the
+        // existing sleds, plus Crucible zones on all pools.  So if we ensure
+        // all these zones exist, we should see no change.
+        for (sled_id, sled_resources) in &policy.sleds {
+            builder.sled_ensure_zone_internal_ntp(*sled_id).unwrap();
+            for pool_name in &sled_resources.zpools {
+                builder
+                    .sled_ensure_zone_crucible(*sled_id, pool_name.clone())
+                    .unwrap();
+            }
+        }
+
+        let blueprint2 = builder.build();
+        let diff = blueprint1.diff(&blueprint2);
+        println!(
+            "initial blueprint -> next blueprint (expected no changes):\n{}",
+            diff
+        );
+        assert_eq!(diff.sleds_added().count(), 0);
+        assert_eq!(diff.sleds_removed().count(), 0);
+        assert_eq!(diff.sleds_changed().count(), 0);
+
+        // The next step is adding these zones to a new sled.
+        let new_sled_id = Uuid::new_v4();
+        let _ = policy_add_sled(&mut policy, new_sled_id);
+        let mut builder =
+            BlueprintBuilder::new_based_on(&blueprint2, &policy, "test_basic");
+        builder.sled_ensure_zone_internal_ntp(new_sled_id).unwrap();
+        let new_sled_resources = policy.sleds.get(&new_sled_id).unwrap();
+        for pool_name in &new_sled_resources.zpools {
+            builder
+                .sled_ensure_zone_crucible(new_sled_id, pool_name.clone())
```

Comment:
I'm a tiny bit surprised the builder is okay with "add internal NTP _and_ crucible in one new build", but I guess it's the builder's caller's responsibility to do any kind of dependency sequencing?

Reply:
Yeah, the basic idea is that the builder is responsible for the mechanism while the planner is responsible for the policy.  It's not so black-and-white, admittedly.  The planner won't let you do things that are easy for it to know cannot work (like putting two zones of the same type on the same zpool).  But "Crucible depends on timesync so it depends on internal NTP and so we don't want to do them in one blueprint" is sufficiently high level that I think the planner should be responsible for it.